### ISSUE TITLE
battery_status.on_click() notification not working when battery is full

### ISSUE THREAD

**Issue by m45t3r** [ultrabug/py3status#509]
I never get the string format set in `format_notify_charging` to work, including if my battery is charging or full charged (and the charger still plugged). When my battery is charging I get the ETA to fully charge my battery, when my battery is fully charged I get a "?", that probably comes from https://github.com/ultrabug/py3status/blob/master/py3status/modules/battery_level.py#L199.

I think the problem is in https://github.com/ultrabug/py3status/blob/master/py3status/modules/battery_level.py#L143:

```python
        if self.time_remaining:
            format = self.format_notify_discharging
        else:
            format = self.format_notify_charging
```
`self.time_remaining` is never `None` in my system, causing the bug (probably).

JFY, when my battery is full I get this from ACPI:

```
~ » acpi                                                    ruby-2.3.1p112
Battery 0: Full, 100%
```

**Comment by tobes** [comment]
What is you output of `acpi -b -i` as that is the actual command called.

Also what version of py3status are you running

**Comment by m45t3r** [comment]
```
~ » acpi -b -i                                                                                                                          ruby-2.3.1p112
Battery 0: Full, 100%
Battery 0: design capacity 4029 mAh, last full capacity 3556 mAh = 88%
```
```
~ » LANG=C pacman -Qi py3status                                                                                                         ruby-2.3.1p112
Name            : py3status
Version         : 3.2-1
Description     : An extensible i3status replacement/wrapper written in python
Architecture    : any
URL             : http://www.ultrabug.fr/tag/py3status/
Licenses        : custom: Simplified BSD
Groups          : None
Provides        : None
Depends On      : python  python-setuptools
Optional Deps   : i3status: for some of the functionality [installed]
Required By     : py3status-modules
Optional For    : None
Conflicts With  : None
Replaces        : None
Installed Size  : 1248.00 KiB
Packager        : Unknown Packager
Build Date      : Thu Oct 13 07:09:25 2016
Install Date    : Thu Oct 13 07:09:26 2016
Install Reason  : Explicitly installed
Install Script  : No
Validated By    : None
```

**Comment by tobes** [comment]
Ok,  I think that this is just some confusion over the settings (which there are many of and not so clear).

`format_notify_discharging` and `format_notify_charging` are used for notifications when you click on the module. But `notification` must be set to True for this.

`format` is used to display the actual output to the i3bar.

**Comment by m45t3r** [comment]
Yeah, I know:

![2016-10-13-11 48 55-screenshot](https://cloud.githubusercontent.com/assets/844343/19353841/1467c1e0-913b-11e6-8463-55afb7b61902.png)

Look at the notification from dunst. It shows a "?", instead of what I put in py3status config file:

```
battery_level {
        cache_timeout = 5
        battery_id = 0
        blocks = ""
        charging_character = "⚡"
        format = "{icon} {percent}%"
        format_notify_charging = "Charging ({percent}%), finishing in: {time_remaining}"
        format_notify_discharging = "Time remaining: {time_remaining}"
        notification = true
        notify_low_level = true
}
```

**Comment by m45t3r** [comment]
P.S.: Actually, it shows ```Time remaining: ?```. It only showed ```?``` because I removed ```format_notify_*``` to do a test.

**Comment by tobes** [comment]
> Actually, it shows Time remaining: ?

good that makes it less perplexing

**Comment by tobes** [comment]
could you provide the output of `acpi -b -i` when the battery is charging/ discharging.

It looks like you are getting an IndexError from the parsing of this.  This module really could do with a cleanup.

**Comment by m45t3r** [comment]
Discharging:

```
~ » acpi -b -i                                              ruby-2.3.1p112
Battery 0: Discharging, 99%, 03:08:05 remaining
Battery 0: design capacity 4029 mAh, last full capacity 3556 mAh = 88%
```
```
Charging almost fully:
```
~ » acpi -b -i                                              ruby-2.3.1p112
Battery 0: Unknown, 96%
Battery 0: design capacity 4029 mAh, last full capacity 3556 mAh = 88%
```
Charging:
```
~ » acpi -b -i                                              ruby-2.3.1p112
Battery 0: Charging, 29%, 01:19:26 until charged
Battery 0: design capacity 4029 mAh, last full capacity 3529 mAh = 87%
```

**Comment by tobes** [comment]
Is it possible for you to apply this patch and see what we get in the logs (when charging and discharging) when the battery is full this will just cause noise in the logs.

```
diff --git a/py3status/modules/battery_level.py b/py3status/modules/battery_level.py
index 55761e9..f851aa4 100644
--- a/py3status/modules/battery_level.py
+++ b/py3status/modules/battery_level.py
@@ -196,6 +196,7 @@ class Py3status:
                 "(?<=, )(\d+:\d+:\d+)(?= remaining)|"
                 "(?<=, )(\d+:\d+:\d+)(?= until)", acpi_battery_lines[0])[0])
         except IndexError:
+            self.py3.log(acpi_battery_lines)
             battery["time_remaining"] = '?'
 
         return battery
```

**Comment by m45t3r** [comment]
```
2016-10-14 09:21:42 INFO i3status spawned using config file /tmp/py3status_p9ny4gud
2016-10-14 09:21:42 INFO loading module "sysdata" from py3status.modules.sysdata
2016-10-14 09:21:42 INFO loading module "battery_level" from py3status.modules.battery_level
2016-10-14 09:21:42 INFO loading module "net_rate" from py3status.modules.net_rate
2016-10-14 09:21:42 INFO loading module "keyboard_layout" from py3status.modules.keyboard_layout
2016-10-14 09:21:42 INFO loading module "arch_updates" from py3status.modules.arch_updates
2016-10-14 09:21:42 INFO loading module "xrandr" from py3status.modules.xrandr
2016-10-14 09:21:46 INFO py3status: Time remaining: 01:29:26
2016-10-14 09:21:46 INFO py3status: Time remaining: 01:21:11
2016-10-14 09:22:58 INFO py3status: Time remaining: 01:25:23
```

**Comment by m45t3r** [comment]
Just one bizarre thing: I compiled from git to do this test (before I was using 3.2 release), and instead of getting notification I got this:

![2016-10-14-09 27 22-screenshot](https://cloud.githubusercontent.com/assets/844343/19387433/83eda470-91f0-11e6-8292-a1ad26799dd9.png)

Going back to 3.2 fixed this.

**Comment by tobes** [comment]
I'm pretty sure that this whole issue is around translations but the logging you posted has not had any output from the patch as yet.  It will only occur when charging/discharging.

If I'm correct then it should be possible to fix your problem.

**Comment by m45t3r** [comment]
Btw, I understand that you want to test the case my battery is full (that is when I get the `Time remaining: ?` from the Exception). However, there is still the problem that when my battery is charging, py3status is using `format_notify_discharging` instead of `format_notify_charging`.

As you can see in my config file posted before, `format_notify_charging = "Charging ({percent}%), finishing in: {time_remaining}"`, however I get `Time remaining: {time_remaining}` instead, that is my `format_notify_discharging`.

**Comment by m45t3r** [comment]
Anyway, wouldn't it be better to rewrite this plugin using `/sys/class/power_supply/*` like i3status itself does? It provides simple files so you wouldn't need to parse it using regex.

I can probably rewrite the module to use it, it would be a nice weekend project.

**Comment by tobes** [comment]
>  However, there is still the problem that when my battery is charging,

that is why I asked 'Is it possible for you to apply this patch and see what we get in the logs (when charging and discharging)'

I need to confirm the actual output seen by py3status as I believe it may contain some Portuguese.

You will need to run the patch for a while to get some output it should look like

2016-10-14 13:51:29 INFO Module `battery_level`: ['Battery 0: Discharging, 99%, 02:24:13 remaining', 'Battery 0: design capacity 6028 mAh, last full capacity 3666 mAh = 60%']

**Comment by tobes** [comment]
> wouldn't it be better to rewrite this plugin using /sys/class/power_supply/*

This maybe true but getting the time to charge then becomes an issue.  Interestingly dbus give hugely different guesses of times than acpi does I'm not sure why.

Also should we also support /proc/.. for old systems?

**Comment by m45t3r** [comment]
> This maybe true but getting the time to charge then becomes an issue. Interestingly dbus give hugely different guesses of times than acpi does I'm not sure why.

Good point. Can you link me the way to do this with dbus?

> Also should we also support /proc/.. for old systems?

@tobes Well, I would prefer to limit myself to `/sys/class/power_supply/*`. This is what i3 uses, and this is what my system support so it is easier to test. If someone needs `/proc` support, I think they can figure it out theirselves.

Unless of course, the format is described somewhere and I can try to implement blindly. However, very rare something implemently without test works.

**Comment by tobes** [comment]
if you want to go down the /sys/class/power_supply/* route then that's great but I'd make it an option so we can give it some testing etc but use acpi too.

The main issue is that the module needs a massive cleanup before adding different ways to get the data is easy to implement.

Anyhow I think the following patch might fix the issue for you even though I'm debugging this blind

```
diff --git a/py3status/modules/battery_level.py b/py3status/modules
/battery_level.py
index 55761e9..36a42a8 100644
--- a/py3status/modules/battery_level.py
+++ b/py3status/modules/battery_level.py
@@ -84,6 +84,7 @@ from __future__ import division  # python2 compat
ibility
 from time import time
 from re import findall
 
+import os
 import math
 import subprocess
 
@@ -120,6 +121,8 @@ class Py3status:
     show_percent_with_blocks = None
 
     def __init__(self):
+        self.env = os.environ.copy()
+        self.env['CL_ALL'] = 'C'
         self.last_known_status = ''
 
     def battery_level(self):
@@ -227,6 +230,7 @@ class Py3status:
         '''
         acpi_raw = subprocess.check_output(
             ["acpi", "-b", "-i"],
+            env=self.env,
             stderr=subprocess.STDOUT)
 
         acpi_list = acpi_raw.decode("UTF-8").split('\n')
```

**Comment by m45t3r** [comment]
Maybe do something like this?

https://stackoverflow.com/questions/26888636/how-to-calculate-the-time-remaining-until-the-end-of-the-battery-charge

Probably wouldn't be precise, however it is a ETA anyway.

**Comment by tobes** [comment]
That would work but initially would be crazy (needs time to get samples etc)

I'd be interested if the patch works because if so then it will fix other modules for non-english users when we are doing dirty parsing from the command line.  I'm not really able to test this locally.

**Comment by m45t3r** [comment]
@tobes Tried the patch, still getting `Time remaining` when my charger is plugged. I don't think it is a LOCALE issue, acpi client does not even have translations: https://sourceforge.net/p/acpiclient/code/ci/master/tree/

**Comment by tobes** [comment]
Ok but without the logs I can't fix this.  The issue is about the parsing and I need to see what is failing.

The clicking stuff is just getting in the way of the problem. If you set `format = "{icon} {percent}% {time_remaining}"` do you __ever__ see the time remaining?

**Comment by m45t3r** [comment]
Yeah.

**Comment by tobes** [comment]
> Yeah.

Ok, it is probably the acpi giving `charging at zero rate - will never fully charge.` response.  Is this issue happening just after you connect/disconnect power?

**Comment by m45t3r** [comment]
Here we go what happens when my battery is full:

```
2016-10-14 11:28:18 INFO Module `battery_level`: ['Battery 0: Full, 100%', 'Battery 0: design capacity 4029 mAh, last full capacity 3540 mAh = 87%']
2016-10-14 11:28:23 INFO Module `battery_level`: ['Battery 0: Full, 100%', 'Battery 0: design capacity 4029 mAh, last full capacity 3540 mAh = 87%']
2016-10-14 11:28:28 INFO Module `battery_level`: ['Battery 0: Full, 100%', 'Battery 0: design capacity 4029 mAh, last full capacity 3540 mAh = 87%']
2016-10-14 11:28:33 INFO Module `battery_level`: ['Battery 0: Full, 100%', 'Battery 0: design capacity 4029 mAh, last full capacity 3540 mAh = 87%']
2016-10-14 11:28:34 INFO py3status: Time remaining: ?
2016-10-14 11:28:34 INFO Module `battery_level`: ['Battery 0: Full, 100%', 'Battery 0: design capacity 4029 mAh, last full capacity 3540 mAh = 87%']
2016-10-14 11:28:39 INFO Module `battery_level`: ['Battery 0: Full, 100%', 'Battery 0: design capacity 4029 mAh, last full capacity 3540 mAh = 87%']
```

**Comment by m45t3r** [comment]
BTW, looking at acpi source code:

```C
		} else if (!strcasecmp(state, "charging")) {
		    if (present_rate > MIN_PRESENT_RATE) {
			seconds = 3600 * (last_capacity - remaining_capacity) / present_rate;
			poststr = " until charged";
		    } else {
			poststr = "charging at zero rate - will never fully charge.";
			seconds = -1;
		    }
		} else if (!strcasecmp(state, "discharging")) {
		    if (present_rate > MIN_PRESENT_RATE) {
			seconds = 3600 * remaining_capacity / present_rate;
			poststr = " remaining";
		    } else {
			poststr = "discharging at zero rate - will never fully discharge.";
			seconds = -1;
		    }
```

The code is quite trivial and does not seem to consider past measurements. So implement something similar in pure Python is trivial.

**Comment by m45t3r** [comment]
Well, found the problem. More details in PR #512.

Anyway, I did end the port of battery_status to sys module. It works. The work can be founded in #511.

**Comment by tobes** [comment]
fixed in #512
